cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -93,39 +94,69 @@ private void upgradeSystemDatasets() throws Exception {
-                                                              supportsIncrement);
-        }
-
-        @Override
-        protected boolean upgradeTable(HTableDescriptor tableDescriptor) {
-          // we don't do any other changes apart from coprocessors upgrade
-          return false;
-        }
-
-        @Override
-        public void create() throws IOException {
-          // no-op
-          throw new UnsupportedOperationException("This DatasetAdmin is only used for upgrade() operation");
-        }
-      };
-      admin.upgrade();
-      LOG.info("Upgraded hbase table: {}", tableId);
+    for (HTableDescriptor desc : hAdmin.listTables()) {
+      String tableName = desc.getNameAsString();
+      if (!isTableInSystemNamespace(tableName) &&
+           isUserTable(tableName) &&
+           isTableCreatedByCDAP(desc)) {
+        upgradeUserTable(desc);
+      }
     }
   }
 
+
+  private void upgradeUserTable(HTableDescriptor desc) throws IOException {
+    HTableNameConverter hTableNameConverter = new HTableNameConverterFactory().get();
+    TableId tableId = hTableNameConverter.from(desc);
+    LOG.info("Upgrading hbase table: {}, desc: {}", tableId, desc);
+
+    final boolean supportsIncrement = HBaseTableAdmin.supportsReadlessIncrements(desc);
+    final boolean transactional = HBaseTableAdmin.isTransactional(desc);
+    DatasetAdmin admin = new AbstractHBaseDataSetAdmin(tableId, hConf, hBaseTableUtil) {
+      @Override
+      protected CoprocessorJar createCoprocessorJar() throws IOException {
+        return HBaseTableAdmin.createCoprocessorJarInternal(cConf,
+                                                            locationFactory,
+                                                            hBaseTableUtil,
+                                                            transactional,
+                                                            supportsIncrement);
+      }
+
+      @Override
+      protected boolean upgradeTable(HTableDescriptor tableDescriptor) {
+        return false;
+      }
+
+      @Override
+      public void create() throws IOException {
+        // no-op
+        throw new UnsupportedOperationException("This DatasetAdmin is only used for upgrade() operation");
+      }
+    };
+    admin.upgrade();
+    LOG.info("Upgraded hbase table: {}", tableId);
+  }
+
+  // Note: This check can be safely used for user table since we create meta.
+  // CDAP-2963 should be fixed so that we can make use of this check generically for all cdap tables
+  private boolean isTableCreatedByCDAP(HTableDescriptor desc) {
+    return (desc.getValue("cdap.version") != null);
+  }
+
+  private boolean isTableInSystemNamespace(String tableName) {
+    // dataset in system namespace starts with <prefix>_system ex: cdap_system
+    return tableName.startsWith(String.format("%s_%s", this.datasetTablePrefix, "system"));
+  }
+
+  private boolean isUserTable(String tableName) {
```

Comment:
This method depends on being used in conjunction with ``isTableInSystemNamespace``, right - meaning you have to call ``!isTableInSystemNamespace(tableName) && isUserTable(tableName)``? 

Perhaps change name/add doc to reflect that?

Or maybe just call this method ``isCDAPUserTable``, and do the ``isTableInSystemNamespace`` and ``isTableCreatedByCDAP`` checks inside this method, so users only have to call this method.